streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/bolt/query/RealtimeJoinBolt.java

```diff
@@ -0,0 +1,638 @@
+import java.util.Map;
+import java.util.Arrays;
+
+
+public class RealtimeJoinBolt extends BaseRichBolt  {
+    private static final Logger LOG = LoggerFactory.getLogger(RealtimeJoinBolt.class);
+    final static String EVENT_PREFIX = StreamlineEvent.STREAMLINE_EVENT + ".";
+
+    private String dataStream;
+    private String lookupStream;
+
+    protected FieldSelector[] outputFields = null;  // specified via bolt.select() ... used in declaring Output fields
+    private String outputStream;
+    private int retentionTime;
+    private int retentionCount;
+    private boolean timeBasedRetention;
+
+    private LinkedListMultimap<String, TupleInfo> lookupBuffer;
+
+    private ArrayDeque<Long> timeTracker; // for time based retention
+    private ArrayList<JoinInfo> joinCriteria = new ArrayList<>();
+
+    private OutputCollector collector;
+    private boolean dropOlderDuplicates;
+    private boolean streamLineProjection = false; // NOTE: Streamline Specific
+
+
+    protected enum JoinType {INNER, LEFT, RIGHT, OUTER}
+    private JoinType joinType;
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        String[] outputFieldNames = new String[outputFields.length];
+        for( int i=0; i<outputFields.length; ++i ) {
+            outputFieldNames[i] = outputFields[i].getOutputName() ;
+        }
+        if (outputStream !=null) {
+            declarer.declareStream(outputStream, new Fields(outputFieldNames));
+        } else {
+            declarer.declare(new Fields(outputFieldNames));
+        }
+    }
+
+    @Override
+    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+        this.collector = collector;
+        if (timeBasedRetention) {
+            lookupBuffer =  LinkedListMultimap.create(50_000);
+            timeTracker = new ArrayDeque<Long>();
+        } else { // count Based Retention
+            lookupBuffer = LinkedListMultimap.create(retentionCount);
+        }
+    }
+
+    // Use streamId, source component name OR field in tuple to distinguish incoming tuple streams
+    public enum Selector { STREAM, SOURCE }
+    protected final Selector selectorType;
+
+    /**
+     * Constructor
+     * @param streamType   Specifies whether 'dataStream' refers to a stream name or source component id.
+     */
+    public RealtimeJoinBolt(Selector streamType) {
+        selectorType = streamType;
+    }
+
+
+    // NOTE: streamline specific
+    /**
+     * Calls  RealtimeJoinBolt(Selector.STREAM)
+     */
+    public RealtimeJoinBolt() {
+        this(Selector.STREAM);
+    }
+
+    /**
+     * Specify Fields to use for join. Field name can be nested x.y.z  (assumes x & y are of type Map<> )
+     * @param dataStreamField    Field to use for join on data stream.
+     * @param lookupStreamField  Field to use for join on lookup stream.
+     * @return
+     */
+    public RealtimeJoinBolt equal(String dataStreamField, String lookupStreamField) {
```

Comment:
hmm .. i personally prefer equal over equiOnFields. Let me think and see if i can come up with an alternative name.